### PR TITLE
Display random profile on homepage

### LINKED ISSUE #7575: [FEATURE] Display random profile on homepage
### Description

1. api to get random profiles (list of users from environment variable)
2. display randomly selected profile from list on the homepage

note: use code from here to created UI https://github.com/EddieHubCommunity/LinkFree/pull/6998/files#diff-bd5cb53f7f62529ac08cd24699132a88e8567b73d93b5289764cfa29442b5720R139-R197
Screenshot 2023-06-14 at 14 32 54

### Screenshots

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by samthatcode**
fixes #7575

<a href="https://gitpod.io/#https://github.com/EddieHubCommunity/LinkFree/pull/7617"></a>

### CODE REVIEW

**Comment by SaraJaoude** [comment]
@samthatcode there is a PR template to help you fill in the right items.  Please update the description with what issue this PR relates to.

**Comment by Vishrut19** [comment]
I have added some changes but I am not able to push it to this PR. How to do it?

**Comment by amandamartin-dev** [comment]
Hey @samthatcode responding here to your question in discord.  To get more information when you see a build failure you will want to click on "Details" and then you will be able to scroll and see information about the errors.

In your case you are failing some accessibility tests that appear to have to do with color contrast. 

Take a look at the full errors and see if they give you some clues where to start. If you still need help, you can ask in Discord in contributor-support and be sure to link the PR so folks can have the context.

**Comment by amandamartin-dev** [comment]
> > Hey @samthatcode responding here to your question in discord.  To get more information when you see a build failure you will want to click on "Details" and then you will be able to scroll and see information about the errors.
> > In your case you are failing some accessibility tests that appear to have to do with color contrast.
> > Take a look at the full errors and see if they give you some clues where to start. If you still need help, you can ask in Discord in contributor-support or help and be sure to link the PR so folks can have the context.
> 
> Alright thanks. Will do that

One thing to try when you are running your code is if you are ever adding something to the UI, make sure to check in dark and light mode.  If you added anything new, that's likely the culprit here.

File: pages/api/randomProfile.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
+// pages/api/randomProfile.js
```
Comment (eddiejaoude) [inline pages/api/randomProfile.js:1]: Please move filter to `api/discover/random.js`

File: pages/api/randomProfile.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+require('dotenv').config();
+
+
+const { MongoClient } = require('mongodb');
+
+// console.log('MONGODB_URI:', process.env.MONGODB_URI);
+
+
+const uri = process.env.MONGODB_URI;
```
Comment (eddiejaoude) [inline pages/api/randomProfile.js:10]: Please see other files on how to do this, we do not use `process`

File: pages/api/randomProfile.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+const client = new MongoClient(uri, { useNewUrlParser: true, useUnifiedTopology: true });
```
Comment (eddiejaoude) [inline pages/api/randomProfile.js:11]: Not sure why you are creating another client here, we have one available to use already, please see other api files

File: pages/api/randomProfile.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    const database = client.db('linkfree');
+
+    // Use the "profiles" collection
+    const collection = database.collection('profiles');
+    const userProfiles = await collection.find({}).toArray();
+    // console.log(userProfiles); // Add this line to log the retrieved profiles
+
+
+    // Select a random profile from the userProfiles array
+    const randomProfile = userProfiles[Math.floor(Math.random() * userProfiles.length)];
```
Comment (eddiejaoude) [inline pages/api/randomProfile.js:29]: Mongo can do the random selection, rather than selecting fetching all the profiles and doing it in js

File: pages/index.js [hunk 5]
```diff
@@ -41,6 +44,38 @@ export default function Home({ total, today }) {
   const router = useRouter();
   const newsletter = router.query.newsletter;
 
+  const [randomProfile, setRandomProfile] = useState(null);
+  const [loading, setLoading] = useState(true);
+
+  useEffect(() => {
```
Comment (eddiejaoude) [inline pages/index.js:50]: This would be better in `getStaticProps` similar to how the page does the other stats - then it will benefit from the caching

**Review by eddiejaoude — CHANGES_REQUESTED** [review]
Thank you for the contribution 👍 

There are too many changes not related to the Issue/PR, please undo those changes

File: package.json [hunk 6]
```diff
@@ -76,7 +76,7 @@
   "devDependencies": {
     "@axe-core/playwright": "^4.7.1",
     "@babel/core": "^7.21.8",
-    "@playwright/test": "^1.34.1",
+    "@playwright/test": "^1.35.1",
```
Comment (eddiejaoude) [inline package.json:79]: This should be in another PR as I don't think it is required here? It is better to keep PRs focused

**Comment by bharathkalyans** [comment]
Is this issue still open?

**Comment by eddiejaoude** [comment]
> Is this issue still open?

@bharathkalyans this is a pull request not an issue

File: pages/api/randomProfile.js (hunk 1 above)
Comment (samthatcode) [reply pages/api/randomProfile.js:1]: Done!

File: pages/api/randomProfile.js (hunk 2 above)
Comment (samthatcode) [reply pages/api/randomProfile.js:10]: Done!

File: pages/api/randomProfile.js (hunk 3 above)
Comment (samthatcode) [reply pages/api/randomProfile.js:11]: Modified and used the readily available

File: pages/api/randomProfile.js (hunk 4 above)
Comment (samthatcode) [reply pages/api/randomProfile.js:29]: Implemented!

File: pages/index.js (hunk 5 above)
Comment (samthatcode) [reply pages/index.js:50]: I'm having a question,  Since the random profile should change on every page reload, using getStaticProps would mean that the random profile would remain the same until the next build, which is not the desired behavior. Correct? Or it should not change?.

Using the useEffect hook to fetch the random profile on the client-side when the component mounts. This way, we will get a new random profile on each page reload, as intended.? Since we are populating the homepage with a random Profile!

I'm gonna use the getStaticProps regardless. Just my Opinion or what i thought it should be. Thanks and i will be expecting feedback on this one

Comment (samthatcode) [reply pages/index.js:50]: > This would be better in `getStaticProps` similar to how the page does the other stats - then it will benefit from the caching

I'm having a question,  Since the random profile should change on every page reload, using getStaticProps would mean that the random profile would remain the same until the next build, which is not the desired behavior. Correct? Or it should not change?.

Using the useEffect hook to fetch the random profile on the client-side when the component mounts. This way, we will get a new random profile on each page reload, as intended.? Since we are populating the homepage with a random Profile!

I'm gonna use the getStaticProps regardless. Just my Opinion or what i thought it should be. Thanks and i will be expecting feedback on this one

Comment (eddiejaoude) [reply pages/index.js:50]: To keep the page performant it will be cache with the other stats, so it will change depending on the config

```js
  return {
    props: { total: totalStats, today: todayStats },
    revalidate: pageConfig.revalidateSeconds,
  };
```

Comment (samthatcode) [reply pages/index.js:50]: > To keep the page performant it will be cache with the other stats, so it will change depending on the config
> 
> ```js
>   return {
>     props: { total: totalStats, today: todayStats },
>     revalidate: pageConfig.revalidateSeconds,
>   };
> ```

Yes understood, done!

**Comment by samthatcode** [comment]
> Thank you for the contribution 👍 
> 
> There are too many changes not related to the Issue/PR, please undo those changes

Hello Sir, I recently checked the code and it was @Vishrut19 that made those changes while he was still assigned to the issue. When I first Open a PR and saw that the tests failed, it was then I messaged him and asked for clarifications and he assured me that he could fix it since we were both assigned to the issue, but then he made those changes and added comment of Test Passing but when I merged the code to my branch, the test didn't actually pass and I saw that he had made changes to existing code. What do you suggest I do about the merged code that resulted in the changes? He repositioned the className mostly. Kindly get back to me to know what to do sir. Also, I have made all the changes required by your comments and that's why I marked them resolved with comments. Waiting to push and to get your feedback first. Thanks for the heads up and teachings

**Comment by Vishrut19** [comment]
> > Thank you for the contribution 👍
> > There are too many changes not related to the Issue/PR, please undo those changes
> 
> Hello Sir, I recently checked the code and it was @Vishrut19 that made those changes while he was still assigned to the issue. When I first Open a PR and saw that the tests failed, it was then I messaged him and asked for clarifications and he assured me that he could fix it since we were both assigned to the issue, but then he made those changes and added comment of Test Passing but when I merged the code to my branch, the test didn't actually pass and I saw that he had made changes to existing code. What do you suggest I do about the merged code that resulted in the changes? He repositioned the className mostly. Kindly get back to me to know what to do sir. Also, I have made all the changes required by your comments and that's why I marked them resolved with comments. Waiting to push and to get your feedback first. Thanks for the heads up and teachings

For me, the tests were passing. Later idk.

**Comment by eddiejaoude** [comment]
no problem @samthatcode , don't worry, we can take a look together over the next few days and we will figure it out 👍

**Comment by samthatcode** [comment]
I have pushed the changes required and I think the tests didn't throw any error. However, I noticed the build failing 1. I'm not sure I know what that means. Please I need clarification on it.  I hope you can review the code to see the changes and give me feedback on it. Thanks for the opportunity

**Comment by amandamartin-dev** [comment]
> I have pushed the changes required and I think the tests didn't throw any error. However, I noticed the build failing 1. I'm not sure I know what that means. Please I need clarification on it. I hope you can review the code to see the changes and give me feedback on it. Thanks for the opportunity

Hey @samthatcode! Thanks for keeping on this. I'm not sure about that error specifically, but I noticed you bumped the version of playwright in your package.json.  I would remove these changes from your PR and run again first in case the version change conflicted with our current test flow.

File: pages/index.js [hunk 7]
```diff
@@ -210,7 +220,7 @@ export default function Home({ total, today }) {
     <>
       <PageHead />
 
-      <div className="bg-primary-low dark:drop-shadow-none dark:bg-primary-high mb-8 p-8 drop-shadow-md">
+      <div className="p-8 mb-8 bg-primary-low dark:drop-shadow-none dark:bg-primary-high drop-shadow-md">
```
Comment (eddiejaoude) [inline pages/index.js:223]: These changes look the same but some classes have been moved around? This makes it difficult to review, please undo changes like this

File: components/user/UserProfile.js [hunk 8]
```diff
@@ -33,6 +34,10 @@ function UserProfile({ BASE_URL, data }) {
     </Link>
   );
 
+  if (!data) {
+    return <p className="text-3xl text-primary-low">No profile data</p>;
+  }
```
Comment (eddiejaoude) [inline components/user/UserProfile.js:39]: This check should be done outside the component, so the component is not rendered

File: pages/index.js [hunk 9]
```diff
@@ -24,20 +24,30 @@ import GitHubAccelerator from "@components/GitHubAccelerator";
...
+
 
 export async function getStaticProps() {
   const pageConfig = config.isr.homepage; // Fetch the specific configuration for this page
 
   const { stats: totalStats } = await getTotalStats();
   const { stats: todayStats } = await getTodayStats();
 
+  // Fetch random Profiles
+  const response = await fetch(`${process.env.NEXT_PUBLIC_BASE_URL}/api/discover/random`);
```
Comment (eddiejaoude) [inline pages/index.js:37]: We do not use `process` but `zod` please check other files for example usage

File: pages/index.js [hunk 10]
```diff
@@ -222,14 +232,17 @@ export default function Home({ total, today }) {
...
           <span className="text-4xl font-bold text-secondary-high dark:text-secondary-low">
             LinkFree
           </span>
           <span className="text-2xl dark:text-primary-low text-primary-medium">
             100% Open Source
           </span>
         </h2>
+        <div className="mt-10 md:mt-20">
+          <UserProfile data={data} />
+        </div>
```
Comment (eddiejaoude) [inline pages/index.js:245]: These margins will be displayed when there is no profile to show, I suggest checking for data first (please rename to `randomProfile`)


```suggestion
        {data && <div className="mt-10 md:mt-20">
          <UserProfile data={data} />
        </div>}
```

**Comment by eddiejaoude** [comment]
@samthatcode your PR is almost there, let me know if you would like to discuss anything (in EddieHub Discord might be better)

File: components/user/UserProfile.js (hunk 8 above)
Comment (samthatcode) [reply components/user/UserProfile.js:39]: Done!

File: pages/index.js (hunk 10 above)
Comment (samthatcode) [reply pages/index.js:245]: Done!

File: pages/index.js (hunk 9 above)
Comment (samthatcode) [reply pages/index.js:37]: Done!

File: pages/index.js (hunk 7 above)
Comment (samthatcode) [reply pages/index.js]: I have undo the changes back to how it was originally. I hope it the review easier

**Comment by eddiejaoude** [comment]
I will resolve conflicts in temporary branch
